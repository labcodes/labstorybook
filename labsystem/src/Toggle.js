import React from "react";
import PropTypes, { oneOf } from "prop-types";
import { isUndefined } from "lodash";

export default class Toggle extends React.Component {
  static propTypes = {
    color: oneOf(["teal", "purple"]),
    disabled: PropTypes.bool,
    defaultValue: PropTypes.bool,
    value: PropTypes.bool,
    name: PropTypes.string.isRequired,
    handleToggle: PropTypes.func,
  };

  static defaultProps = {
    color: "teal",
    disabled: false,
    defaultValue: false,
    value: undefined,
    handleToggle: undefined,
  };

  constructor(props) {
    super(props);
    const { defaultValue, value, name } = props;
    if (!isUndefined(defaultValue) && !isUndefined(value)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both value and defaultValue for input ${name} at the same time. We always initialize the toggle with defaultValue. Make sure this is the behaviour you want.`
      );
    }
    this.state = {
      localValue: defaultValue,
    };
  }

  handleOnChange = (e) => {
    const { handleToggle } = this.props;
    if (!isUndefined(handleToggle)) {
      handleToggle(e);
    }

    this.setState((state) => ({ localValue: !state.localValue }));
  };

  render() {
    const { color, name, disabled, value } = this.props;
    const { localValue } = this.state;
    const isChecked = !isUndefined(value) ? value : localValue;

    return (
      <label
        className={`
          lab-toggle
          ${color ? `lab-toggle--${color}` : ""}
          ${disabled ? " lab-toggle--disabled" : ""}
        `}
        htmlFor={name}
      >
        <input
          type="checkbox"
          id={name}
          className="lab-toggle__input"
          checked={isChecked}
          onChange={this.handleOnChange}
          {...(disabled ? { disabled } : undefined)}
        />
        <span className="lab-toggle__slider" />
      </label>
    );
  }
}
