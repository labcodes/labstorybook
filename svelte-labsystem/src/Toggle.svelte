<script>
  export let color = undefined;
  export let disabled = undefined;
  export let value = false;
  export let name;
</script>

<label
  class={`lab-toggle${color ? ` lab-toggle--${color}` : ''}`}
  class:lab-toggle--disabled={disabled}
  htmlFor={name}>
  <input
    type="checkbox"
    id={name}
    class="lab-toggle__input"
    on:change
    bind:checked={value}
    {name}
    {disabled} />
  <span class="lab-toggle__slider" />
</label>
