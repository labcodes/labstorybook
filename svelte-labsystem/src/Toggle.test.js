import { render } from "@testing-library/svelte";
import Toggle from "./Toggle.svelte";

describe("Toggle", () => {
  it("renders with base props", async () => {
    const { container } = render(Toggle, {
      name: "test",
    });
    expect(container).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const { container } = render(Toggle, {
      name: "test",
      disabled: true,
    });

    expect(container).toMatchSnapshot();
    expect(container.querySelector("input").disabled).toBeTruthy();
  });

  it("renders as expected when passing a different theme", async () => {
    const { container } = render(Toggle, {
      name: "test",
      theme: "purple",
    });
    expect(container).toMatchSnapshot();
  });
});
