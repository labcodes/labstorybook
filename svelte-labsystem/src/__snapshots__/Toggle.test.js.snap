// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders as expected when passing a different theme 1`] = `
<body>
  <div>
    <label
      class="lab-toggle"
      for="test"
    >
      <input
        class="lab-toggle__input"
        id="test"
        name="test"
        type="checkbox"
      />
       
      <span
        class="lab-toggle__slider"
      />
    </label>
  </div>
</body>
`;

exports[`Toggle renders as expected when passing disabled as true 1`] = `
<body>
  <div>
    <label
      class="lab-toggle lab-toggle--disabled"
      for="test"
    >
      <input
        class="lab-toggle__input"
        disabled=""
        id="test"
        name="test"
        type="checkbox"
      />
       
      <span
        class="lab-toggle__slider"
      />
    </label>
  </div>
</body>
`;

exports[`Toggle renders with base props 1`] = `
<body>
  <div>
    <label
      class="lab-toggle"
      for="test"
    >
      <input
        class="lab-toggle__input"
        id="test"
        name="test"
        type="checkbox"
      />
       
      <span
        class="lab-toggle__slider"
      />
    </label>
  </div>
</body>
`;
